src/rcc/clocking.rs

```diff
@@ -0,0 +1,292 @@
+//! knowledge is encoded in the SysClkSources enum.
+//!
+//! Each enum variant contains information about the clock it represents. Some clocks are
+//! configurable, and thus have fields, and some are not. For example, the LSI
+//! (LowSpeedInternalRC) clock is always 32 kHz, but the MSI (MediumSpeedInternalRC) clock can
+//! be configured and thus has a frequency component.
+//!
+//! To use them, compose them and feed them to, e.g., sysclk.
+//!
+//! ```rust
+//! let mut rcc = RCC.constrain();
+//! let msi_clk = clocking::MediumSpeedInternalRC::new(8_000_000, false);
+//! let sys_clk_src = clocking::SysClkSource::MSI(msi_clk);
+//! let cfgr = rcc.cfgr.sysclk(sys_clk_src);
+//! ```
+//!
+//! The PLL is a bit more complex because it _is_ a source (`PLLClkOutput`) and also _requires_
+//! a source (`PLLClkSource`), but you compose the types similarly.
+
+use super::rcc;
+
+pub trait InputClock {
+    fn freq(&self) -> u32;
+}
+
+/// High-speed internal 16 MHz RC
+#[derive(Clone, Copy)]
+pub struct HighSpeedInternal16RC {
+    pub always_on: bool,
+    pub auto_start: bool,
+}
+
+impl HighSpeedInternal16RC {
+    /// Applies the selection options to the configuration registers and turns the clock on
+    pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+        rcc.cr.modify(|_, w| {
+            w.hsion()
+                .set_bit()
+                .hsikeron()
+                .bit(self.always_on)
+                .hsiasfs()
+                .bit(self.auto_start)
+        });
+        while rcc.cr.read().hsirdy().bit_is_clear() {}
+        (16_000_000, 0b01)
+    }
+}
+
+/// Medium-speed internal 100 kHz - 48 MHz RC
+#[derive(Clone, Copy)]
+pub struct MediumSpeedInternalRC {
+    freq: u32,
+    auto_cal: bool,
+}
+
+impl MediumSpeedInternalRC {
+    pub fn new(freq: u32, auto_cal: bool) -> Self {
+        MediumSpeedInternalRC { freq, auto_cal }
+    }
+
+    pub fn bits(&self) -> u8 {
+        match self.freq {
+            100_000 => 0b0000,
+            200_000 => 0b0001,
+            400_000 => 0b0010,
+            800_000 => 0b0011,
+            1_000_000 => 0b0100,
+            2_000_000 => 0b0101,
+            4_000_000 => 0b0110,
+            8_000_000 => 0b0111,
+            16_000_000 => 0b1000,
+            24_000_000 => 0b1001,
+            32_000_000 => 0b1010,
+            48_000_000 => 0b1011,
+            _ => panic!("bad MSI speed value!"),
+        }
+    }
+
+    /// Configures the MSI to the specified frequency, and enables hardware
+    /// auto-calibration if requested by enabling (and waiting for) the LSE.
+    pub fn configure(&self, rcc: &rcc::RegisterBlock) -> (u32, u8) {
+        rcc.cr
+            .modify(|_, w| unsafe { w.msirange().bits(self.bits()).msirgsel().set_bit() });
+        while rcc.cr.read().msirdy().bit_is_clear() {}
+
+        if self.auto_cal {
+            // FIXME This... may not work? I'm not sure if I've got a board problem or using
```

Comment:
Is this still an issue?

If LSE is not ready then mot likely you failed to enable it.
You can try to check register's value after you enable it.
The only reason I can think of, is that BDCR write protection is not lifted...